docs/enhancements/fep-000-api-device-fleet.md

```diff
@@ -0,0 +1,434 @@
+configuration to the fleet mandated configuration and remove any modifications made to the
+device spec.
+
+Devices set to `spec.managementMode: manual`, but with a `metadata.owner` fleet , must signal
+a device status condition to indicate that the device is not in sync with the fleet because of
+this.
+
+```yaml
+  - lastTransitionTime: "2024-04-30T15:06:20Z"
+    message: This device is not in sync with the fleet
+    reason: Manually Managed
+    status: False
+    type: Management
+```
+
+When a fleet owned device is moved back to fleet management the status condition should be removed.
+
+##### `/device/{name}` with managementMode: manual
+
+Devices in manual management mode are not managed by the fleet controller, the device
+controller will not attempt to reconcile the device spec with the fleet spec.
+
+In this mode, the device.spec is fully writtable through this API.
+
+If the device is selected by a fleet, the `spec.managementMode` can be flipped to `fleet`, and the
+device controller will start managing the device again.
+
+While devices are managed by fleets, all floating tags are resolved to stable hashes, but
+when a device is managed manually, the floating tags can be used, see the next sub-section
+for additional details.
+
+###### Expected behaviour regarding external references with floating tags
+
+Changes on external references is monitored and acted on (git branches,
+device images, secrets), the device controller should re-evaluate the device
+spec and apply the changes to the rendered version which will be exposed
+on the `/device/{name}/rendered` endpoint.
+
+Several mechanisms can be used to track external references:
+* Polling
+* Webhooks
+
+#### `/device/{name}/rendered` (device side)
+This endpoint is used by the device agent to retrieve the rendered device spec, this endpoint is
+read-only, each device can only access it's own device rendered spec based on the device
+certificate identification, an administrator can access any device rendered spec for debug purposes.
+
+The agent can pass two parameters: `version`and `owner` if those haven't changed the API will return
+204 No Content, if the version or fleet has changed the API will return the rendered spec.
+
+Change proposed:
+* The parameter name `fleet` is changed to `owner` to be consistent with the metadata field.
+
+Behavior changes:
+* If a device configuration is not yet rendered for some reason, FC will queue the device for rendering and instruct
+  the device to try again later by returning 202 Accepted.
+
+The administrator can find the fully rendered device spec by using the following cmdline command:
+```shell
+flightctl get device/xxxxxxx --rendered
+```
+
+The JSON blob configuration exposed to the device in the /device/{name}/rendered endpoint
+while compacted for transport, should be formated in a way that it's easy to read by humans
+when printed out to the console, by adding tabs and newlines.
```

Comment:
I'd say that non-formatted may be useful, and those who want it formatted should pipe the output to jq.

Reply:
But it's hard to pass to jq, because it's embedded in yaml.

What would be the use case? When consumed by the user, this API is only expected to be used for debugging.

We could add a flag not to prettify output, but IMHO, the default should be targeting the main use case (debugging by 
a user).

Reply:
Today, GET /api/v1/devices/{name}/specification returns JSON.

Reply:
It's not plain json today, only the config field.

```
./bin/flightctl debug devspec f07526ed0c3c251d8764c83833f5135824375c06f46e3d3feb080fc29875a4c3
config: '{"ignition":{"config":{"replace":{"verification":{}}},"proxy":{},"security":{"tls":{}},"timeouts":{},"version":"3.4.0"},"kernelArguments":{},"passwd":{},"storage":{"files":[{"group":{},"overwrite":true,"path":"/etc/motd","user":{},"contents":{"source":"data:,This%20system%20is%20managed%20by%20flightctl.%0A","verification":{}},"mode":422},{"group":{},"overwrite":true,"path":"/etc/microshift/manifests/agent-registration.yaml","user":{"name":"root"},"contents":{"source":"data:text/plain;charset=utf-8;base64,Ci0tLQph...10K","verification":{}},"mode":420},{"group":{},"overwrite":true,"path":"/etc/microshift/manifests/deployment.yaml","user":{"name":"root"},"contents":{"source":"data:text/plain;charset=utf-8;base64,YXB...DAK","verification":{}},"mode":420},{"group":{},"overwrite":true,"path":"/etc/microshift/manifests/kustomization.yaml","user":{"name":"root"},"contents":{"source":"data:text/plain;charset=utf-8;base64,cmVzb3VyY2VzOgotIG5hbWVzcGFjZS55YW1sCi0gZGVwbG95bWVudC55YW1sCi0gc2VydmljZS55YW1sCi0gcm91dGUueWFtbAotIGNyZC1rbHVzdGVybGV0LnlhbWwKLSBhZ2VudC1yZWdpc3RyYXRpb24ueWFtbAo=","verification":{}},"mode":420},{"group":{},"overwrite":true,"path":"/etc/microshift/manifests/namespace.yaml","user":{"name":"root"},"contents":{"source":"data:text/plain;charset=utf-8;base64,YXB..==","verification":{}},"mode":420},{"group":{},"overwrite":true,"path":"/etc/microshift/manifests/route.yaml","user":{"name":"root"},"contents":{"source":"data:text/plain;charset=utf-8;base64,YX...mUK","verification":{}},"mode":420},{"group":{},"overwrite":true,"path":"/etc/microshift/manifests/service.yaml","user":{"name":"root"},"contents":{"source":"data:text/plain;charset=utf-8;base64,YXB...MK","verification":{}},"mode":420},{"group":{},"overwrite":true,"path":"/etc/crio/openshift-pull-secret","user":{"name":"root"},"contents":{"source":"data:text/plain;charset=utf-8;base64,.....WlsIjoicm5vcmllZ2FAcmVkaGF0LmNvbSJ9fX0K","verification":{}},"mode":420}]},"systemd":{}}'
containers: {}
os:
  image: quay.io/flightctl/flightctl-agent-basic-nginx:latest
owner: ""
systemd:
  matchPatterns:
  - microshift.service
  - crio.service
  - flightctl-agent.service
templateVersion: "2024-04-24T10:30:17Z"

```

Reply:
You should be able to use the `-o json` flag on the CLI.

Reply:
Ok, I will remove any details about output formating, and I propose an RFE is started for the device side transport when we are ready for that.